src/lib/optimizer/strategy/predicate_placement_rule.cpp

```diff
@@ -41,19 +60,54 @@ void PredicatePlacementRule::_push_down_traversal(const std::shared_ptr<Abstract
+  // A helper method for cases where the input_node does not allow us to proceed.
   const auto handle_barrier = [&]() {
     _insert_nodes(current_node, input_side, push_down_nodes);
 
-    if (input_node->left_input()) {
+    const auto& barrier_node = input_node;
+    const auto barrier_node_is_pushdown_predicate = [&barrier_node]() {
+      if (barrier_node->type == LQPNodeType::Predicate) {
+        const auto predicate_node = std::static_pointer_cast<PredicateNode>(barrier_node);
+        return !_is_expensive_predicate(predicate_node->predicate());
+      } else if (barrier_node->type == LQPNodeType::Join) {
+        const auto join_mode = std::static_pointer_cast<JoinNode>(barrier_node)->join_mode;
+        return join_mode == JoinMode::Semi || join_mode == JoinMode::AntiNullAsTrue ||
+               join_mode == JoinMode::AntiNullAsFalse;
+      }
+      return false;
+    }();
+
+    auto next_push_down_traversal_root = barrier_node;
+    if (barrier_node_is_pushdown_predicate) {
+      // The barrier node is a predicate eligible for pushdown. Therefore, we would like it to be covered by the next
+      // recursion of _push_down_traversal. However, since _push_down_traversal looks at the inputs of a given
+      // root node only, barrier_node would not become pushed down.
+      // To allow for a pushdown of barrier_node, a temporary root node is inserted above it. Afterwards, the temporary
+      // root node is used as a root for the next _push_down_traversal recursion. As a result, barrier_node will be seen
+      // as an input of the temporary root node, and thus become pushed down by the rule, if possible.
+      next_push_down_traversal_root = std::make_shared<TemporaryRootNode>();
+      lqp_insert_node_above(barrier_node, next_push_down_traversal_root);
+    }
+
+    if (next_push_down_traversal_root->left_input()) {
       auto left_push_down_nodes = std::vector<std::shared_ptr<AbstractLQPNode>>{};
-      _push_down_traversal(input_node, LQPInputSide::Left, left_push_down_nodes, estimator);
+      _push_down_traversal(next_push_down_traversal_root, LQPInputSide::Left, left_push_down_nodes, estimator);
+
+      if (next_push_down_traversal_root->right_input()) {
+        auto right_push_down_nodes = std::vector<std::shared_ptr<AbstractLQPNode>>{};
+        _push_down_traversal(next_push_down_traversal_root, LQPInputSide::Right, right_push_down_nodes, estimator);
+      }
     }
-    if (input_node->right_input()) {
-      auto right_push_down_nodes = std::vector<std::shared_ptr<AbstractLQPNode>>{};
-      _push_down_traversal(input_node, LQPInputSide::Right, right_push_down_nodes, estimator);
+
+    // The recursion calls to _push_down_traversal have returned. Therefore, we must remove the temporary root node, we
+    // might have inserted previously. (see comment above)
+    if (std::dynamic_pointer_cast<TemporaryRootNode>(next_push_down_traversal_root)) {
```

Comment:
if we used `LogicalPlanRootNode` above: `next_push_down_traversal_root->type == LQPNodeType::Root && next_push_down_traversal_root->output_count()`